ships/0031-shipwright-trigger.md

```diff
@@ -0,0 +1,296 @@
+
+The [inventory is the central component](https://github.com/otaviof/shipwright-trigger/blob/main/pkg/trigger/inventory/inventory.go) of Shipwright Trigger, it stores all the Build instances organized in a way that allows searching for types of triggers, depending on the Inventory client.
+
+For example, the WebHook Handler will always search for Builds based on the Git repository URL, the type of event (`Push` or `PullRequest`), and the branch names. In other hand, the other Controllers will query the inventory based on the `.objectRef` attribute instead.
+
+As you can see on the diagram above, almost all components are interacting with the Inventory using the specialized query methods `SearchForGit` and `SearchForObjectRef`.
+
+#### Shipwright Build Controller
+
+The Builds are added or removed from the Inventory through the [Build Controller](https://github.com/otaviof/shipwright-trigger/blob/main/pkg/trigger/controllers/build_controller.go), responsible to reflect all Shipwright Build resources into the Inventory. On adding new entries, the Build is prepared for the subsequent queries.
+
+### Tekton Run Controller
+
+Watches for [Tekton Run instances](https://github.com/otaviof/shipwright-trigger/blob/main/pkg/trigger/controllers/run_controller.go) referencing Shipwright Builds, when a new instance is created it creates a new `BuildRun`. The controller also watches over the `BuildRun` instance, in order to reflect the status back to the Tekton Run parent.
+
+The Tekton Run instances are part of the Custom Task workflow, everytime Tekton finds a `TaskRef` resource outside of Tekton's scope, it creates a `Run` instance with the coordinates. In other words, to extend Tekton's functionality third party applications must watch and interact with those objects.
+
+### Tekton PipelineRun Controller
+
+The [controller for PipelineRun instances](https://github.com/otaviof/shipwright-trigger/blob/main/pkg/trigger/controllers/pipelinerun_controller.go) is meant to react when a Pipeline reaches the desired status, so upon changes on the resource the controller checks on the inventory if there are triggers configured for the specific resource in question, in the desired status.
+
+Upon the creation of a `BuildRun` instance, the `PipelineRun` object is labeled for the controller to be able to avoid reprocessing.
+
+### WebHook Handler
+
+The [WebHook handler](https://github.com/otaviof/shipwright-trigger/blob/main/pkg/trigger/webhooks/http_handler.go) is a simple HTTP server implementation which receives requests from the outside, and after processing the event, searches over Builds that should be activated. The search on the inventory happens in the same fashion as the controllers, however uses `SearchForGit` method.
+
+This type of `SearchForGit` is meant to match the repository URL, the type of event and the branches affected. For instance, the WebHook event can have different types, like `Push` or `PullRequest` and plus the branch affected.
+
```

Comment:
Implementation-wise, one question regarding HA. We here have a component with a mixture of a controller (Tekton Run) and a web endpoint (GitHub webhook). In a deployment with more than one replica, what do you envision?

- for the controller, having a leader like in build
- the webhook requests handled by all pods like in a typical web server deployment ?

Reply:
Regarding the HA perspective, it should be the same than a regular Kubernetes controller, the client itself negotiates the active instance. And for the WebHook endpoint, the Ingress Controller will distribute the requests on the healthy pods, and the same is valid for the OpenShift Routes.

> - for the controller, having a leader like in build
> - the webhook requests handled by all pods like in a typical web server deployment ?

In other words, yes for both questions :-)